File: WolframFourierArtCode/FourierFuncs/python/FourierFunc.py
from math import sin;
from math import pi as π;

class Coordinates:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y


class FourierFuncBase:

    _series = None

    min_t = None
    max_t = None
        
    def __init__(self):
        if type(self) is FourierFuncBase:
            raise Exception('FourierFuncBase is an abstract class and cannot be instantiated directly')
        self._init_series()
        self.min_t = self._series[-1].min_t
        self.max_t = self._series[ 0].max_t

    def get_series(self):
        return self._series


    def calc_normal(self, t: float) -> Coordinates:
        denorm_t = self._map_range(t, 0.0, 1.0, self.min_t, self.max_t)

        return self.calc(denorm_t)


    def calc(self, t: float) -> Coordinates:
        series = next((x for x in self._series if x.min_t <= t <= x.max_t), None)
        if series == None:
            return None

        return series.calc(t)


    def _init_series(self):
        return NotImplementedError


    def _map_range(self, value, source_min, source_max, target_min, target_max):
        scale = (target_max - target_min) / (source_max - source_min)
        return target_min + (value - source_min) * scale



class FourierSeriesBase:

    min_t = None
    max_t = None

    def __init__(self):
        if type(self) is FourierSeriesBase:
            raise Exception('FourierSeriesBase is an abstract class and cannot be instantiated directly')

    def calc_normal(self, t: float) -> Coordinates:
        denorm_t = self._scale_argument(t)
        return self.calc(denorm_t)

    def calc(self, t: float) -> Coordinates:
        return NotImplementedError
        
    def _scale_argument(self, t):
        return (self.min_t + t * (self.max_t - self.min_t))

File: WolframFourierArtCode/FourierFuncs/csharp/FourierFunc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace WolframFourierArtCode.FourierFuncs;

/// <summary>
/// Point coordinates
/// </summary>
public class Coordinates(double x, double y)
{
    public double X { get; set; } = x;
    public double Y { get; set; } = y;
}

public interface IFourierFunc: IFourierSeries
{
    /// <summary>
    /// Get all series
    /// </summary>
    IFourierSeries[] GetSeries();
}

public interface IFourierSeries
{
    /// <summary>
    /// Min argument value
    /// </summary>
    double MinT { get; }

    /// <summary>
    /// Max argument value
    /// </summary>
    double MaxT { get; }

    /// <summary>
    /// Calculate coordinates by normalized agrument <paramref name="t"/>.
    /// Argument must belong to [0.0, 1.0] range.
    /// </summary>
    Coordinates CalcNormal(double t);

    /// <summary>
    /// Calculate coordinates by agrument <paramref name="t"/>.
    /// Argument must belong to [<see cref="MinT"/>, <see cref="MaxT"/>] range.
    /// </summary>
    Coordinates Calc(double t);
}

public abstract class FourierFuncBase : IFourierFunc
{
    protected IFourierSeries[] _series = null!;

    protected FourierFuncBase()
    {
        InitSeries();
    }

    public double MinT => _series[^1].MinT;

    public double MaxT => _series[ 0].MaxT;

    public IFourierSeries[] GetSeries() => _series;

    public Coordinates CalcNormal(double t)
    {
        var denormT = MapRange(t, 0d, 1d, MinT, MaxT);

        return Calc(denormT);
    }

    public Coordinates Calc(double t)
    {
        var series = _series
            .Where(s => s.MinT <= t && t <= s.MaxT)
            .FirstOrDefault();

        return series?.Calc(t);
    }

    protected abstract void InitSeries();

    private static double MapRange(
        double value,
        double sourceMin, double sourceMax,
        double targetMin, double targetMax
        )
    {
        var scale = (targetMax - targetMin) / (sourceMax - sourceMin);
        return targetMin + (value - sourceMin) * scale;
    }
}


public abstract class FourierSeriesBase: IFourierSeries
{
    protected double sin(double a) => Math.Sin(a);

    protected const double π = Math.PI;

    public abstract double MinT { get; }

    public abstract double MaxT { get; }

    public Coordinates CalcNormal(double t)
    {
        var scaledT = ScaleArgument(t);
        return Calc(scaledT);
    }

    public abstract Coordinates Calc(double t);

    protected double ScaleArgument(double t) => (MinT + t * (MaxT - MinT));
}
